### PR TITLE
Await for all BinderServerTransport sessions to terminate before retu…

### PR DESCRIPTION
**Pull request by mateusazis**
…rning security policy executor to the object pool.

Addresses #10897.

### CODE REVIEW

File: binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+  /**
+   * @throws IllegalStateException if {@link #scheduleTerminationNotification()} has already been
+   * called.
+   */
+  ServerTransportListener track(
+      ServerTransportListener listener,
+      BinderTransport.BinderServerTransport binderServerTransport) {
+    synchronized (this) {
+      checkState(
```
Comment (jdcormie) [inline binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java:43]: So I believe this can actually happen today: A SETUP_TRANSPORT binder transaction can arrive while some other thread is calling BinderServer#shutdown(). Both shutdown() and handleTransaction() are synchronized so the latter method call on the binder thread will block. The implementation of BinderServer#shutdown() calls hostServiceBinder.detach() to drop any new transactions but the SETUP_TRANSPORT one in-progress has already started being handled. After shutdown() returns, handleTransaction() runs and would throw ISE here.

Instead, can you change BinderServer#handleTransaction() to handle this race by simply ignoring the SETUP_TRANSPORT transaction `if (shutdown)`? This could be a small pre-PR.

File: binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+/**
+ * Tracks which {@link BinderTransport.BinderServerTransport} are currently active and allows
+ * invoking a {@link BinderTransportSecurity.TerminationListener} only once all transports are
+ * terminated.
+ */
+final class ActiveTransportTracker {
+  @GuardedBy("this")
+  private final HashSet<BinderTransport.BinderServerTransport> activeTransports = new HashSet<>();
+  private final BinderTransportSecurity.TerminationListener transportSecurityTerminationListener;
```
Comment (jdcormie) [inline binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java:20]: There's nothing security specific about this class (which is good) so can you make this just a `Runnable` and call it `terminationListener` ?

File: binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  @GuardedBy("this")
+  private boolean awaitingNoActiveTransports = false;
+  @GuardedBy("this")
+  private boolean terminationNotificationRequested = false;
```
Comment (jdcormie) [inline binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java:24]: Can you call this `shutdown` instead? I think it would be a lot clearer.

File: binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+ * Tracks which {@link BinderTransport.BinderServerTransport} are currently active and allows
+ * invoking a {@link BinderTransportSecurity.TerminationListener} only once all transports are
+ * terminated.
+ */
+final class ActiveTransportTracker {
+  @GuardedBy("this")
+  private final HashSet<BinderTransport.BinderServerTransport> activeTransports = new HashSet<>();
+  private final BinderTransportSecurity.TerminationListener transportSecurityTerminationListener;
+  @GuardedBy("this")
+  private boolean awaitingNoActiveTransports = false;
```
Comment (jdcormie) [inline binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java:22]: Do you need this ? If you prohibit tracking new transports when `shutdown` and assume that callers obey the contract of `ServerTransportListener#transportTerminated` is just `shutdown` enough?

File: binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+import io.grpc.internal.ServerTransportListener;
+
+/**
+ * Tracks which {@link BinderTransport.BinderServerTransport} are currently active and allows
+ * invoking a {@link BinderTransportSecurity.TerminationListener} only once all transports are
+ * terminated.
+ */
+final class ActiveTransportTracker {
+  @GuardedBy("this")
+  private final HashSet<BinderTransport.BinderServerTransport> activeTransports = new HashSet<>();
```
Comment (jdcormie) [inline binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java:19]: What are the pros/cons of tracking this as a set vs just counting the number of transports that have been created but not yet terminated?

File: binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java (hunk 1 above)
Comment (mateusazis) [reply binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java]: Ack. Will start working on a separate PR in parallel.

File: binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+  }
+
+  /**
+   * Schedules this tracker to notify the {@link BinderTransportSecurity.TerminationListener}
+   * provided in the constructor once all remaining active transports have terminated. The
+   * notification may happen immediately if there are no active transports.
+   *
+   * <p>There is no guarantee about which thread the TerminationListener will be called from.
+   */
+  void scheduleTerminationNotification() {
```
[the remark this answers is not in the document]
Comment (mateusazis) [reply binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java]: Done. I still kept the `untrack()` method though as I also had to decorate ` ServerTransportListener`.

File: binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java (hunk 2 above)
Comment (mateusazis) [reply binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java]: I originally did it this way because I thought it might be a good idea to hint why the termination listener was needed, since it wasn't used for anything else. I don't feel super strongly about it though. Updated.

File: binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java (hunk 3 above)
Comment (mateusazis) [reply binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java]: Updated.

File: binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java (hunk 4 above)
Comment (mateusazis) [reply binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java]: I think I had indeed added a separate field to make the class 100% safe from misuse, but it may be overkill indeed. Updated.

File: binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java (hunk 5 above)
Comment (mateusazis) [reply binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java]: Using a set would guarantee that whatever transport is terminating is indeed getting tracked. That would protect this class against mishandling:

- the termination of a transport that was never tracked to begin with
- duplicate termination notifications for the same transport

It might be just my paranoia though. Changed to tracking just the count of active transports for now.

File: binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+final class ActiveTransportTracker implements ServerListener  {
+  private final ServerListener delegate;
+  private final Runnable terminationListener;
+  @GuardedBy("this")
+  private boolean shutdown = false;
+  @GuardedBy("this")
+  private int activeTransportCount = 0;
+
+  /**
+   * @param terminationListener invoked only once the server has started shutdown
```
Comment (jdcormie) [inline binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java:26]: Can you javadoc the other @param as well?

File: binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+  }
+
+  @Override
+  public void serverShutdown() {
+    synchronized (this) {
+      shutdown = true;
+
+      if (activeTransportCount == 0) {
+        // We can run immediately
+        terminationListener.run();
```
Comment (jdcormie) [inline binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java:61]: Consider invoking this callback after releasing the synchronized lock, just in case

File: binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+   * @param terminationListener invoked only once the server has started shutdown
+   * ({@link #serverShutdown()} AND the last active transport is terminated.
+   */
+  ActiveTransportTracker(ServerListener delegate, Runnable terminationListener) {
+    this.delegate = delegate;
+    this.terminationListener = terminationListener;
+  }
+
+  @Override
+  public ServerTransportListener transportCreated(ServerTransport transport) {
```
Comment (jdcormie) [inline binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java:35]: I believe this method should be synchronized too, since transportCreated() and untrack() both load/increment/store the counter value and they could be called concurrently by different threads

File: binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java (hunk 8 above)
Comment (jdcormie) [reply binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java:61]: You may want to factor this new more elaborate logic out into a method like `maybeInvokeTerminationCallback()`

File: binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    checkState(
+        !shutdown,
+        "Attempting to track a new BinderServerTransport, but the server shutdown has already" +
```
[the remark this answers is not in the document]
Comment (mateusazis) [reply binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java]: Updated.

File: binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java (hunk 7 above)
Comment (mateusazis) [reply binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java:28]: Updated.

File: binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+  public void serverShutdown() {
+    synchronized (this) {
+      shutdown = true;
+
+      if (activeTransportCount == 0) {
+        // We can run immediately
+        terminationListener.run();
+      }
+    }
+    delegate.serverShutdown();
```
[the remark this answers is not in the document]
Comment (mateusazis) [reply binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java]: Makes sense. Updated.

File: binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java (hunk 8 above)
Comment (mateusazis) [reply binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java]: Probably doesn't matter since this class shouldn't be interacted with after the termination callback is invoked, but agreed in principle that it may be a good idea to release locks earlier if possible.

Updated.

File: binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java (hunk 9 above)
Comment (mateusazis) [reply binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java:37]: Updated.

*/me looks suspiciously at `@GuardedBy`...*

File: binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java (hunk 8 above)
Comment (jdcormie) [reply binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java]: Thanks but it now looks like every call to maybeInvokeTerminationCallback() comes from an already synchronized block so I don't think anything has changed...

For me, the important principle is that, in general, holding a lock while invoking an arbitrary user callback risks deadlock. One DRY approach could be:
```
@GuardedBy("this")
private Runnable getListenerIfTerminated() {
  return (shutdown && activeTransportCount == 0) ? terminationListener : null;
}

private void untrack() {
  Runnable runMeUnlocked;    
  synchronized (this) {
    --activeTransportCount;
    runMeUnlocked = getListenerIfTerminated();
  }
  if (runMeUnlocked != null) { runMeUnlocked.run(); }
}

@Override
public void serverShutdown() {
  delegate.serverShutdown();
  Runnable runMeUnlocked;    
  synchronized (this) {
    shutdown = true;
    runMeUnlocked = getListenerIfTerminated();
  }
  if (runMeUnlocked != null) { runMeUnlocked.run(); }
}
```

OTOH you have a point that no calls after termination are allowed, per ServerListener contract, so feel free to make that case in a comment instead.

Comment (mateusazis) [reply binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java]: Reworked it based on your suggestion. Even if the contract should be stable, holding a lock while some unknown code runs feels like something to avoid regardless.

File: binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    @Override
+    public Attributes transportReady(Attributes attributes) {
+      return delegate.transportReady(attributes);
+    }
+
+    @Override
+    public void transportTerminated() {
+      untrack();
```
Comment (jdcormie) [inline binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java:104]: I wonder if the order should be reversed here too. In other words, we shouldn't declare a server terminated until all its transports are also terminated.

File: binder/src/test/java/io/grpc/binder/internal/ActiveTransportTrackerTest.java [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+
+  @Test
+  public void testTransportCreation_afterServerShutdown_throws() {
+    tracker.serverShutdown();
+
+    assertThrows(IllegalStateException.class, this::registerNewTransport);
+  }
+
+  private ServerTransportListener registerNewTransport() {
+    BinderTransport.BinderServerTransport transport =
```
Comment (jdcormie) [inline binder/src/test/java/io/grpc/binder/internal/ActiveTransportTrackerTest.java:105]: Could this be a fake or mock io.grpc.internal.ServerTransport instead? If so, you wouldn't need to come up with all the deps needed to create a real BinderServerTransport.

File: binder/src/test/java/io/grpc/binder/internal/ActiveTransportTrackerTest.java [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+      new FixedObjectPool<>(Executors.newSingleThreadScheduledExecutor());
+  private final IBinder callbackBinder = new Binder();
+
+  private boolean serverShutdownNotified = false;
+  private final Runnable testShutdownListener = () -> serverShutdownNotified = true;
+  private ActiveTransportTracker tracker;
+
+  @Before
+  public void setUp() {
+    tracker = new ActiveTransportTracker(new NoOpServerListener(), testShutdownListener);
```
Comment (jdcormie) [inline binder/src/test/java/io/grpc/binder/internal/ActiveTransportTrackerTest.java:57]: Can you add test cases that ensure the methods of io.grpc.internal.ServerListener are forwarded to the delegate?Maybe use a mock instead of this noop impl?

**Review by jdcormie — APPROVED** [review]

File: binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java (hunk 1 above)
Comment (jdcormie) [reply binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java]: Cool. Let's get that one it before merging this so we don't risk spurious ISEs

File: binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java (hunk 12 above)
Comment (mateusazis) [reply binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java]: Agreed. Updated.

File: binder/src/test/java/io/grpc/binder/internal/ActiveTransportTrackerTest.java (hunk 14 above)
Comment (mateusazis) [reply binder/src/test/java/io/grpc/binder/internal/ActiveTransportTrackerTest.java]: Added.

File: binder/src/test/java/io/grpc/binder/internal/ActiveTransportTrackerTest.java (hunk 13 above)
Comment (mateusazis) [reply binder/src/test/java/io/grpc/binder/internal/ActiveTransportTrackerTest.java]: There's some internal guidance for preferring real objects over test doubles. I think the mocks are fine for the cases where we need to verify interaction and we don't want to track that ourselves, but this objects seemed relatively test-friendly.

Kept it for now. Let me know if you feel strongly about it.

File: binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java (hunk 1 above)
Comment (mateusazis) [reply binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java]: Tackling it in #11253.

File: binder/src/test/java/io/grpc/binder/internal/ActiveTransportTrackerTest.java (hunk 13 above)
Comment (jdcormie) [reply binder/src/test/java/io/grpc/binder/internal/ActiveTransportTrackerTest.java]: Agreed that in general we prefer to test state rather than interactions (go/tott/281). But in this case, the real BinderServerTransport doesn't expose the state you need in order to test ActiveTransportTracker's function as a `ServerTransportListener` decorator (See go/choose-test-double#interaction-verification-appropriate case reason 2 and 3) In particular, you need to verify that, in addition to the SUT's tracking function, every ServerTransportListener method call results in exactly one method call on the same thread with the same arguments to the delegate ServerTransportListener before returning. BinderServerTransport doesn't record or expose any of the state you'd need to verify this (why would it). OTOH, mocks are perfect for testing decorators like this.

LMK if you buy this.

**Comment by jdcormie** [comment]
LGTM but let's merge #11253 first

**Comment by jdcormie** [comment]
#11260 supersedes #11253 as prerequisite mentioned above.
